target_chains/solana/programs/pyth_solana_receiver/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+};
+
+declare_id!("DvPfMBZJJwKgJsv2WJA8bFwUMn8nFd5Xpioc6foC3rse");
+pub const POST_VAA: u8 = 2;
+
+#[program]
+pub mod pyth_solana_receiver {
+    use {
+        super::*,
+        pythnet_sdk::wire::v1::{
+            AccumulatorUpdateData,
+            MerklePriceUpdate,
+        },
+        serde_wormhole::RawMessage,
+        solana_program::program::invoke,
+        std::hash::Hash,
+        wormhole_sdk::vaa::{
+            Body,
+            Header,
+        },
+    };
+
+    pub fn decode_posted_vaa(ctx: Context<DecodePostedVaa>) -> Result<()> {
+        let posted_vaa = &ctx.accounts.posted_vaa.payload;
+        let batch: BatchPriceAttestation =
+            BatchPriceAttestation::deserialize(posted_vaa.as_slice())
+                .map_err(|_| ReceiverError::DeserializeVAAFailed)?;
+
+        msg!(
+            "There are {} attestations in this batch.",
+            batch.price_attestations.len()
+        );
+
+        for attestation in batch.price_attestations {
+            msg!("product_id: {}", attestation.product_id);
+            msg!("price_id: {}", attestation.price_id);
+            msg!("price: {}", attestation.price);
+            msg!("conf: {}", attestation.conf);
+            msg!("ema_price: {}", attestation.ema_price);
+            msg!("ema_conf: {}", attestation.ema_conf);
+            msg!("num_publishers: {}", attestation.num_publishers);
+            msg!("publish_time: {}", attestation.publish_time);
+            msg!("attestation_time: {}", attestation.attestation_time);
+        }
+
+        Ok(())
+    }
+
+    pub fn update(
+        _ctx: Context<Update>,
+        data: Vec<u8>,
+        recovery_id: u8,
+        signature: [u8; 64],
+    ) -> Result<()> {
+        // This costs about 10k compute units
+        let message_hash = {
+            let mut hasher = keccak::Hasher::default();
+            hasher.hash(&data);
+            hasher.result()
+        };
+
+        // This costs about 25k compute units
+        let recovered_pubkey = secp256k1_recover(&message_hash.0, recovery_id, &signature)
+            .map_err(|_| ProgramError::InvalidArgument)?;
+
+        msg!(
+            "Recovered key: {}",
+            recovered_pubkey.0.encode_hex::<String>()
+        );
+
+        // TODO: Check the pubkey is an expected value.
+        // Here we are checking the secp256k1 pubkey against a known authorized pubkey.
+        //
+        // if recovered_pubkey.0 != AUTHORIZED_PUBLIC_KEY {
+        //  return Err(ProgramError::InvalidArgument);
+        // }
+
+        Ok(())
+    }
+
+    /// Verifies the accumulator update data header then invokes a CPI call to wormhole::postVAA
+    ///
+    /// * `data` - Bytes of the AccumulatorUpdateData response from hermes with the updates omitted
+    ///           (i.e. the `updates` field is an empty array). The updates are removed so that
+    ///           all the data needed for postVaa can fit in one txn.
+    pub fn post_accumulator_update_vaa(
```

Comment:
I think adding some extra types and structs would help to separate the boundary between the pythnet wireformat and the solana receiver wireformat.

Reply:
the reason this is a wrapper instead of having the client do `postVaa` directly was so that we could do on-chain verification of the header in the `accumulatorUpdateData` using the `AccumulatorUpdateData::try_from_slice`

the client/cli code is currently doing the exact same thing when it first parses the accumulatorUpdateData bytes before it invokes this ix but i thought it would be good to have this verification on-chain as well

Reply:
I think the header verification check isn't adding any value here. Someone could circumvent this check by calling post vaa themselves, and then calling the subsequent update method.

All of our security checks need to happen in the `update` method themselves. In this particular case, I don't think the magic number / version checks are security critical so they could potentially be skipped for solana (given that there's a sort of weird flow needed to do the vaa posting).